contracts/javascore/aggregator/src/main/java/relay/aggregator/RelayAggregator.java

```diff
@@ -0,0 +1,305 @@
+            sigs[i] = signatureList.get(i);
+        }
+        return sigs;
+    }
+
+    protected void setSignature(String pktID, Address addr, byte[] sign) {
+        signatures.at(pktID).set(addr, sign);
+    }
+
+    protected static byte[] serializeSignatures(byte[][] sigs) {
+        ByteArrayObjectWriter w = Context.newByteArrayObjectWriter("RLPn");
+        w.beginList(sigs.length);
+
+        for (byte[] sig : sigs) {
+            w.write(sig);
+        }
+
+        w.end();
+        return w.toByteArray();
+    }
+
+    protected static byte[][] deserializeSignatures(byte[] encodedSigs) {
+        ObjectReader r = Context.newByteArrayObjectReader("RLPn", encodedSigs);
+
+        ArrayList<byte[]> sigList = new ArrayList<>();
+
+        r.beginList();
+        while (r.hasNext()) {
+            sigList.add(r.readByteArray());
+        }
+        r.end();
+
+        byte[][] sigs = new byte[sigList.size()][];
+        for (int i = 0; i < sigList.size(); i++) {
+            sigs[i] = sigList.get(i);
+        }
+
+        return sigs;
+    }
+
+    private void adminOnly() {
+        Context.require(Context.getCaller().equals(admin.get()), "Unauthorized: caller is not the leader relayer");
+    }
+
+    private void relayersOnly() {
+        Address caller = Context.getCaller();
+        Boolean isRelayer = false;
+        for (int i = 0; i < relayers.size(); i++) {
+            Address relayer = relayers.get(i);
+            if (relayer.equals(caller)) {
+                isRelayer = true;
+                break;
+            }
+        }
+        Context.require(isRelayer, "Unauthorized: caller is not a registered relayer");
+    }
+
+    private Boolean signatureThresholdReached(String pktID) {
+        int noOfSignatures = 0;
+        for (int i = 0; i < relayers.size(); i++) {
+            Address relayer = relayers.get(i);
+            byte[] relayerSign = signatures.at(pktID).get(relayer);
+            if (relayerSign != null) {
+                noOfSignatures++;
+            }
+        }
+        return noOfSignatures >= signatureThreshold.get();
+    }
+
+    private void removePacket(String pktID) {
+        packets.set(pktID, null);
+        DictDB<Address, byte[]> signDict = signatures.at(pktID);
+
+        for (int i = 0; i < relayers.size(); i++) {
+            Address relayer = relayers.get(i);
+            signDict.set(relayer, null);
+        }
+    }
+
+    @EventLog(indexed = 2)
+    public void PacketRegistered(
+            String srcNetwork,
+            String contractAddress,
+            BigInteger srcSn,
+            BigInteger srcHeight,
+            String dstNetwork,
+            byte[] data) {
+    }
+
+    @EventLog(indexed = 2)
+    public void PacketAcknowledged(
+            String srcNetwork,
+            String contractAddress,
+            BigInteger srcSn,
+            BigInteger srcHeight,
+            String dstNetwork,
+            byte[] data,
+            byte[] signatures) {
+    }
```

Comment:
im not sure if we even need packet acknowledged . we would rather need event for threshold met though.

Reply:
PacketAcknowledged event serves the both purpose of threshold being met and publishing required information for the relayer to perform subsequent operations.